_posts/2018-07-23-performance-profiling-with-craft-cms.md

```diff
@@ -0,0 +1,117 @@
+![Enable Craft Toolbar](http://res.cloudinary.com/dsteinel/image/upload/c_scale,w_512/v1532351329/Espi/blog/performance%20profiling/enable-debugger-toolbar.gif)
+
+### Debugging your page
+For debugging the page-speed, the most important information in the debugger toolbar are: 
+<br>
+* **Time**. Shows you how long the website takes to load 
+* **DB**. Shows how many Database queries and the time they need
+
+If you click anywhere in the toolbar, you can see all the details.
+**Sort by duration** is the main feature we are using here. As a quick reminder, we want to identify queries, which take too much time and need some improvements. If you open up the **Performance Profiling** window, you can identify which part of the template takes the longest. 
+![Craft Toolbar Performance Profiling](http://res.cloudinary.com/dsteinel/image/upload/c_scale,w_512/v1532351330/Espi/blog/performance%20profiling/perfomance-profiling__debugger-toolbar__pp.png)
+
+If you split up your templates in several different templates, it is easier to determine slow parts in the code. You can see in the image above, that the `global-header`  is part of the `layout` template. So in this case the `layout.twig` template is including the partial `global-header.twig`. You can read more about `includes` in the [Twig Docs](https://twig.symfony.com/doc/2.x/tags/include.html)
+
+If your code is not split up like in the example above or if you need to profile your twig template more in detail, check out the [craft-twigprofiler](https://github.com/nystudio107/craft-twigprofiler). With this tool you can test smaller chunks of your template.
+
+### Reduce database queries with Eager Loading
+To keep the database queries as low as possible, you need to bundle the queries into bigger pieces and fetch as many related datas as you can at once.  In order to tell Craft in advance which data we are going to use, we need to **Eager Load** them.
+With Eager Loading, Craft is able to return related elements. Like this Craft fetches sub-elements in advance instead of doing the same query for every sub-element over and over again.
+
+A relation in craft has a source and a target. The source has a relation field and selects the other element. The target is selected by the source.
+[Relations | Craft 2 Documentation](https://docs.craftcms.com/v2/relations.html#terminology)
+> It’s important to grasp those terms, as they are relevant to the templating side of things
+
+
+In Craft, [elements](https://docs.craftcms.com/v2/plugins/working-with-elements.html) are:  
+* Assets
+* Categories
+* Entries
+* Global Sets
+* Matrix Blocks
+* Tags
+* Users 
+<br><br>
+
+It takes some steps to implement Eager Loading.
+This example shows you how you Eager Load an imageField:
+
+#### No Eager Loading
+{% highlight twig %}
```

Comment:
you might want to format the whole block as a code snippet (with the three backticks)

Reply:
This is actually how you highlight code snippets with jekyll. It also works with backticks though.

Reply:
Uh, right. Forgot that the dev blog runs on Jekyll. Thought it was generic Markdown. My bad 😇